apps/arweave/src/ar_http_iface_middleware.erl

```diff
@@ -504,6 +507,66 @@ handle(<<"POST">>, [<<"block2">>], Req, Pid) ->
 	erlang:put(post_block2, true),
 	post_block(request, {Req, Pid, binary}, erlang:timestamp());
 
+%% Accept a (partial) solution from a pool proxy or a CM node and validate it.
+%%
+%% If the node is a CM exit node and a pool client, send the given solution to
+%% the pool and return an empty JSON object.
+%%
+%% If the node is a pool server, return a JSON object:
+%% {
+%%   "indep_hash": "",
+%%   "status": ""
+%% },
+%% where the status is one of "accepted", "accepted_block", "rejected_bad_poa",
+%% "rejected_wrong_hash", "rejected_bad_vdf", "rejected_mining_address_banned",
+%% "stale", "rejected_vdf_not_found", "rejected_missing_key_file".
+%% If the solution is partial, "indep_hash" string is empty.
+handle(<<"POST">>, [<<"partial_solution">>], Req, Pid) ->
+	case ar_node:is_joined() of
+		true ->
+			handle_post_partial_solution(Req, Pid);
+		false ->
+			not_joined(Req)
+	end;
+
+%% Return the information about up to ?GET_JOBS_COUNT latest VDF steps and a difficulty.
+%%
+%% If the given VDF output is present in the latest 10 VDF steps, return only the steps
+%% strictly above the given output. If the given output is our latest output,
+%% wait for up to ?GET_JOBS_TIMEOUT_S and return an empty list if no new steps are
+%% computed by the time. Also, only return the steps strictly above the latest block.
+%%
+%% If we are a pool server, return the current network difficulty along with the VDF
+%% information.
+%%
+%% If we are a CM exit node and a pool client, return the partial difficulty provided
+%% by the pool.
```

Comment:
For my own understanding  - why do we have CM miners communicate with the pool proxy via the CM exit node? e.g. why don't they query jobs directly from the pool proxy?

Reply:
The motivation is to save traffic.

Reply:
My soft opinion is that this *may* fall under premature optimization in that we're adding complexity to the code to optimize for a scenario that may not be an issue. 

i.e. We're only optimizing the case where many small miners have coordinated into a CM cluster, but we aren't optimizing the case where many independent small miners join the pool directly. It may turn out that neither scenario causes enough traffic to be an issue, or it may turn out that the CM case rarely happens - in both those scenarios the optimization won't help us at all, but we're still left with a lot of code that needs to duplicately handle communicating with an Exit Node as well as communicating directly with the Pool. The cost there is both in terms of increased code complexity, but also the increased surface area to be tested and retested.

I personally would prefer to start simple and have all nodes communicate directly with the pool and only add in "Exit Node as Proxy" if we find a material bottleneck in practice.

Reply:
There could be a case when workers are behind firewall, and cm host:port is actually port-forwarded from other network. In this case worker will have job, but can't post solution

This is kinda edge case, but for China it's kinda possible. Otherwise you need setup VPN for whole cluster

Reply:
Okay I think I understand. Is this accurate:
1. When a CM cluster joins a pool, the Pool only interacts with the Exit Node. So from the perspective of the pool, only a single miner has joined the pool. The pool doesn't know (and doesn't need to know) how many actual miners are running behind the exit node.

2. By the same token the CM miners themselves never interact with the pool host, they only interact with the Exit node so from their perspective, the Exit node is the pool host (they get jobs from the exit node and share partial solutions with the exit node)

Reply:
Logged a backlog issue to see if there's anyway to simplify the duplicate codepaths: https://github.com/ArweaveTeam/arweave-dev/issues/354

but not a priority for this PR.